### PR TITLE
 Add support for child span count

### LINKED ISSUE #306: Tracing: Add child span count to Span.
See https://github.com/census-instrumentation/opencensus-proto/blob/6d79e46232e026136ae013489ef611ca721cb839/src/opencensus/proto/trace/v1/trace.proto#L302.

### PR DESCRIPTION
**Pull request by mayurkale22**
Fixes #306  
This is based on spec: See https://github.com/census-instrumentation/opencensus-proto/blob/6d79e46232e026136ae013489ef611ca721cb839/src/opencensus/proto/trace/v1/trace.proto#L302.

### CODE REVIEW

File: packages/opencensus-core/test/test-root-span.ts [hunk 1]
```diff
@@ -169,6 +184,7 @@ describe('RootSpan', () => {
       for (const span of root.spans) {
         assert.ok(span.ended);
       }
+      assert.equal(root.numberOfChildren, 1);
```
Comment (rghetia) [inline packages/opencensus-core/test/test-root-span.ts]: there is no need to check for child count in other test case unless the test case affects the logic.

File: packages/opencensus-core/test/test-root-span.ts [hunk 2]
```diff
@@ -206,6 +222,7 @@ describe('RootSpan', () => {
       rootSpan.addAnnotation('description test', {} as Attributes, Date.now());
 
       assert.ok(rootSpan.annotations.length > 0);
+      assert.equal(rootSpan.numberOfChildren, 0);
```
Comment (rghetia) [inline packages/opencensus-core/test/test-root-span.ts]: ditto.

File: packages/opencensus-core/test/test-root-span.ts (hunk 1 above)
Comment (mayurkale22) [reply packages/opencensus-core/test/test-root-span.ts]: Sure, Done.

**Comment by mayurkale22** [comment]
> What would you think about making the child count a core property of the `Span` type? That would then enable situations where there is a larger operation root span with sub-operations as child spans and those child spans then make RPC calls.

AFAIK there isn't a notion of grandchild spans in node (all child spans are direct children of a root span). https://github.com/census-instrumentation/opencensus-node/issues/105

**Comment by draffensperger** [comment]
> AFAIK there isn't a notion of grandchild spans in node (all child spans are direct children of a root span). #105

Is that by design based on the specs / other language implementations? Or is that more of a specific limitation of OpenCensus Node at this time?

**Comment by mayurkale22** [comment]
@draffensperger allow me to rephrase my previous sentence https://github.com/census-instrumentation/opencensus-node/pull/332#issuecomment-462457456.

AFAIK in terms of single process or server, there will be only two types of requests (incoming and outgoing). Root spans corresponds to incoming requests and child span for outgoing. There is only in and out (no grand out). So, if there are multiple micro services running we connect all these spans through common traceId (which is propagated through context). 

> What would you think about making the child count a core property of the `Span` type?

I think it makes sense to have child count property of the ```RootSpan``` type, as there is no way to create child from child (nested child).

**Comment by draffensperger** [comment]
I agree that for RPCs you typically only have in and out, but if we allow for other types of spans, I think we can get the grandchild case. I'm thinking of custom intermediate spans that represent larger conceptual things like "process middleware" or "load recommendations" and similar, that each may have multiple RPCs or other custom spans in them. As a user views a trace, they could then see which high level operation in the request is slow and then drill down to the subcalls of that operation. This of course requires that the user sets up those custom spans, but ideally I would think we would let them do that.

**Comment by draffensperger** [comment]
I think the work I'm suggesting is captured well in a previous issue that I just commented on: https://github.com/census-instrumentation/opencensus-node/issues/105#issuecomment-468304948

So I'm fine with merging this and then adding support for nested child spans at a later time.

**Comment by mayurkale22** [comment]
> I think the work I'm suggesting is captured well in a previous issue that I just commented on: [#105 (comment)](https://github.com/census-instrumentation/opencensus-node/issues/105#issuecomment-468304948)
> 
> So I'm fine with merging this and then adding support for nested child spans at a later time.

Ok cool, thanks.

**Pull request merged**